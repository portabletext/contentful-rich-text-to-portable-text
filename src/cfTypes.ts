export interface CFNode {
  nodeType: string
  data: {[key: string]: any}
  content?: CFNode[]
}

export interface CFContainerNode extends CFNode {
  content: CFNode[]
}

export interface CFDocumentNode extends CFContainerNode {
  nodeType: 'document'
}

export interface CFHeadingNode extends CFContainerNode {
  nodeType: 'heading-1' | 'heading-2' | 'heading-3' | 'heading-4' | 'heading-5' | 'heading-6'
}

export interface CFParagraphNode extends CFContainerNode {
  nodeType: 'paragraph'
}

export interface CFTextNode extends CFNode {
  nodeType: 'text'
  value: string
  marks: CFMark[]
}

export interface CFHyperlinkNode extends CFContainerNode {
  nodeType: 'hyperlink'
  data: {uri: string}
}

export interface CFEntryHyperlinkNode extends CFContainerNode {
  nodeType: 'entry-hyperlink'
  data: {target: CFEntryReference}
}

export interface CFAssetHyperlinkNode extends CFContainerNode {
  nodeType: 'asset-hyperlink'
  data: {target: CFAssetReference}
}

export interface CFListItemNode extends CFContainerNode {
  nodeType: 'list-item'
}

export interface CFUnorderedListNode extends CFContainerNode {
  nodeType: 'unordered-list'
  content: CFListItemNode[]
}

export interface CFOrderedListNode extends CFContainerNode {
  nodeType: 'ordered-list'
  content: CFListItemNode[]
}

export interface CFHorizontalRuleNode extends CFContainerNode {
  nodeType: 'hr'
}

export interface CFBlockQuoteNode extends CFContainerNode {
  nodeType: 'blockquote'
}

export interface CFEmbeddedEntryBlockNode extends CFContainerNode {
  nodeType: 'embedded-entry-block'
  data: {
    target: CFEntryReference
  }
}

export interface CFEmbeddedEntryInlineNode extends CFContainerNode {
  nodeType: 'embedded-entry-inline'
  data: {
    target: CFEntryReference
  }
}

export interface CFEmbeddedAssetBlockNode extends CFContainerNode {
  nodeType: 'embedded-asset-block'
  data: {
    target: CFAssetReference
  }
}

export interface CFEntryReference {
  sys: {
    id: string
    type: 'Link'
    linkType: 'Entry'
  }
}

export interface CFAssetReference {
  sys: {
    id: string
    type: 'Link'
    linkType: 'Asset'
  }
}

export type CFReference = CFEntryReference | CFAssetReference

export interface CFMark {
  type: 'bold' | 'italic' | 'underline' | 'code'
}

export function isCFTextNode(node: CFNode): node is CFTextNode {
  return node.nodeType === 'text'
}
