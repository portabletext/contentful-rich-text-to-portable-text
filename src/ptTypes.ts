export interface SanityFieldType extends Record<string, any> {
  name: string
  type: string
  title?: string
}

export interface SanitySchemaType extends Record<string, any> {
  type: 'object' | 'document'
  name: string
  fields: SanityFieldType[]
}

export interface PTNode {
  _type: string
  _key: string
}

export interface PTBreak extends PTNode {
  _type: 'break'
  _key: string
  style: 'lineBreak' | 'readMore'
}

export interface PTSpan extends PTNode {
  _type: 'span'
  _key: string
  marks: string[]
  text: string
}

export interface PTMark extends PTNode {
  _type: string
  _key: string
  [key: string]: any
}

export interface PTBlock extends PTNode {
  _type: 'block'
  style: string
  children: PTNode[]
  markDefs: PTMark[]
  listItem?: string
  level?: number
}

export interface PTListItem extends PTBlock {
  listItem: string
  level: number
}

export interface PTVirtualList extends PTNode {
  _type: 'virtualList'
  type: string
  level: number
  children: PTNode[]
}

export interface PTLink extends PTMark {
  _type: 'link'
  href: string
}

export interface PTReference extends PTNode {
  _type: 'reference'
  _ref: string
  _weak?: boolean
}

export const markMapping: {[key: string]: string} = {
  bold: 'strong',
  italic: 'em'
}

export function isPTBlockNode(node: PTNode): node is PTBlock {
  return node._type === 'block'
}
