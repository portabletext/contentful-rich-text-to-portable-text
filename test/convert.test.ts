import objectHash from 'object-hash'
import {toPortableText, TransformOptions} from '../src'
import {CFNode, CFAssetHyperlinkNode} from '../src/cfTypes'
import heading from './fixtures/heading'
import marks from './fixtures/marks'
import doubleMarks from './fixtures/doubleMarks'
import paragraph from './fixtures/paragraph'
import deepList from './fixtures/deepList'
import link from './fixtures/link'
import list from './fixtures/list'
import allFeatures from './fixtures/allFeatures'
import {PTLink} from 'ptTypes'

const generateKey = (node: CFNode) => `k${objectHash(node).slice(0, 7)}`
const options = {generateKey}

describe('toPortableText', () => {
  it('basic h1', () => {
    const pt = toPortableText(heading, options)
    expect(pt).toHaveLength(1)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'h1',
      markDefs: [],
      children: [{_type: 'span', marks: [], text: 'Hello, everyone!'}]
    })
  })

  it('basic marks', () => {
    const pt = toPortableText(marks, options)
    expect(pt).toHaveLength(1)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'h2',
      markDefs: [],
      children: [
        {_type: 'span', marks: ['strong'], text: 'Hello'},
        {_type: 'span', marks: [], text: ', '},
        {_type: 'span', marks: ['underline'], text: 'everyone!'}
      ]
    })
  })

  it('double marks', () => {
    const pt = toPortableText(doubleMarks, options)
    expect(pt).toHaveLength(1)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'h3',
      markDefs: [],
      children: [
        {_type: 'span', marks: ['strong', 'em'], text: 'Hello'},
        {_type: 'span', marks: [], text: ', '},
        {_type: 'span', marks: ['underline'], text: 'everyone!'}
      ]
    })
  })

  it('paragraph', () => {
    const pt = toPortableText(paragraph, options)
    expect(pt).toHaveLength(1)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          marks: [],
          text: "This is a little introduction. It's just a paragraph without any fancy stuff."
        }
      ]
    })
  })

  it('link', () => {
    const pt = toPortableText(link, options)
    expect(pt).toHaveLength(1)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'normal',
      markDefs: [
        {
          _type: 'link',
          href: 'https://www.techopedia.com/definition/23666/platform-agnostic'
        }
      ],
      children: [
        {
          _type: 'span',
          _key: 'k1864115',
          marks: [],
          text: 'When representing text with formatting (such as '
        },
        {
          _type: 'span',
          marks: ['strong'],
          text: 'bold'
        },
        {
          _type: 'span',
          marks: [],
          text: ', '
        },
        {
          _type: 'span',
          marks: ['em'],
          text: 'italic'
        },
        {
          _type: 'span',
          marks: [],
          text: ', '
        },
        {
          _type: 'span',
          marks: ['underline'],
          text: 'underline'
        },
        {
          _type: 'span',
          marks: [],
          text: ' or '
        },
        {
          _type: 'span',
          marks: ['code'],
          text: 'code'
        },
        {
          _type: 'span',
          marks: [],
          text: '), you want to do this in a '
        },
        {
          _type: 'span',
          marks: ['k933711f'],
          text: 'platform-agnostic'
        },
        {
          _type: 'span',
          marks: [],
          text: ' way, so you are able to render it on any device/surface.'
        }
      ]
    })
  })

  it('unordered list', () => {
    const pt = toPortableText(list, options)
    expect(pt).toHaveLength(2)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'bullet',
      level: 1,
      markDefs: [],
      children: [
        {
          _type: 'span',
          marks: [],
          text: 'Simple lists are supported, of course.'
        }
      ]
    })
    expect(pt[1]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'bullet',
      level: 1,
      markDefs: [],
      children: [
        {
          _type: 'span',
          marks: [],
          text: "It's full of "
        },
        {
          _type: 'span',
          marks: ['strong'],
          text: 'bullets'
        }
      ]
    })
  })

  it('deep lists', () => {
    const pt = toPortableText(deepList, options)
    expect(pt[0]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'bullet',
      level: 1,
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'k6494432',
          marks: [],
          text: 'Simple lists are supported, of course.'
        }
      ]
    })

    expect(pt[1]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'number',
      level: 2,
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'k2a71eca',
          marks: [],
          text: 'Deeper is cooler?'
        }
      ]
    })

    expect(pt[2]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'number',
      level: 2,
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'k3c2b38c',
          marks: ['strong'],
          text: 'Especially with multiple items?'
        }
      ]
    })

    expect(pt[3]).toMatchObject({
      _type: 'block',
      _key: 'k4f770d8',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'k25f5438',
          marks: [],
          text: "Even when they're split"
        }
      ],
      level: 1,
      listItem: 'bullet'
    })

    expect(pt[4]).toMatchObject({
      _type: 'block',
      style: 'normal',
      listItem: 'bullet',
      level: 1,
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'k2005001',
          marks: [],
          text: 'Trailing bullet at level 1.'
        }
      ]
    })
  })

  it('supports all features', () => {
    const pt = toPortableText(allFeatures, options)
    expect(pt).toMatchSnapshot()
  })

  it('can control asset handling', () => {
    const pt = toPortableText(allFeatures, {
      ...options,
      transformers: {
        'asset-hyperlink': (node: CFAssetHyperlinkNode, options: TransformOptions): PTLink[] => {
          const link = options.transformers.link()
          return [
            {
              _type: 'link',
              _key: options.generateKey(node, options),
              href: 'https://foo.bar'
            }
          ]
          //const target = console.log(node, void options)
          //return []
        }
      }
    })
    expect(pt).toMatchSnapshot()
  })
})
