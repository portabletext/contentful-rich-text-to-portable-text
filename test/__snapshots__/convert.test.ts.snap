// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toPortableText can control asset handling 1`] = `
Array [
  Object {
    "_key": "ka3360a9",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kd9865f3",
        "_type": "span",
        "marks": Array [],
        "text": "Hello, everyone!",
      },
    ],
    "markDefs": Array [],
    "style": "h1",
  },
  Object {
    "_key": "kbf477f7",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k9ee64b5",
        "_type": "span",
        "marks": Array [],
        "text": "This is a little introduction. It's just a paragraph without any fancy stuff.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k3ec023b",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k77063cd",
        "_type": "span",
        "marks": Array [],
        "text": "Background",
      },
    ],
    "markDefs": Array [],
    "style": "h2",
  },
  Object {
    "_key": "kf72d303",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k1864115",
        "_type": "span",
        "marks": Array [],
        "text": "When representing text with formatting (such as ",
      },
      Object {
        "_key": "k972fe1e",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "bold",
      },
      Object {
        "_key": "kab7a70e",
        "_type": "span",
        "marks": Array [],
        "text": ", ",
      },
      Object {
        "_key": "ka074963",
        "_type": "span",
        "marks": Array [
          "em",
        ],
        "text": "italic",
      },
      Object {
        "_key": "kab7a70e",
        "_type": "span",
        "marks": Array [],
        "text": ", ",
      },
      Object {
        "_key": "kd3067d7",
        "_type": "span",
        "marks": Array [
          "underline",
        ],
        "text": "underline",
      },
      Object {
        "_key": "kbeb12c2",
        "_type": "span",
        "marks": Array [],
        "text": " or ",
      },
      Object {
        "_key": "ka1bcaba",
        "_type": "span",
        "marks": Array [
          "code",
        ],
        "text": "code",
      },
      Object {
        "_key": "ke530559",
        "_type": "span",
        "marks": Array [],
        "text": "), you want to do this in a ",
      },
      Object {
        "_key": "ka50426c",
        "_type": "span",
        "marks": Array [
          "k933711f",
        ],
        "text": "platform-agnostic",
      },
      Object {
        "_key": "kaebcdd6",
        "_type": "span",
        "marks": Array [],
        "text": " way, so you are able to render it on any device/surface.",
      },
    ],
    "markDefs": Array [
      Object {
        "_key": "k933711f",
        "_type": "link",
        "href": "https://www.techopedia.com/definition/23666/platform-agnostic",
      },
    ],
    "style": "normal",
  },
  Object {
    "_key": "k1380a38",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k4f7db8f",
        "_type": "span",
        "marks": Array [],
        "text": "The situation gets more complex when you want to combine ",
      },
      Object {
        "_key": "kafe874d",
        "_type": "span",
        "marks": Array [
          "strong",
          "em",
          "k1c64ec3",
        ],
        "text": "multiple",
      },
      Object {
        "_key": "kccbade6",
        "_type": "span",
        "marks": Array [],
        "text": " formatting options, or when several \\"marks\\" are split ",
      },
      Object {
        "_key": "k0b7a9d5",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "across ",
      },
      Object {
        "_key": "k88a89d3",
        "_type": "span",
        "marks": Array [
          "strong",
          "underline",
        ],
        "text": "multiple",
      },
      Object {
        "_key": "k9e33411",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": " words",
      },
      Object {
        "_key": "k1159efe",
        "_type": "span",
        "marks": Array [],
        "text": ".",
      },
    ],
    "markDefs": Array [
      Object {
        "_key": "k1c64ec3",
        "_type": "link",
        "href": "https://multiple.com/",
      },
    ],
    "style": "normal",
  },
  Object {
    "_key": "k9c4bf48",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k6494432",
        "_type": "span",
        "marks": Array [],
        "text": "Simple lists are supported, of course.",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kdff0bd7",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kb19e8e0",
        "_type": "span",
        "marks": Array [],
        "text": "Indented ones as well",
      },
    ],
    "level": 2,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k4606371",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k99acc14",
        "_type": "span",
        "marks": Array [],
        "text": "But you can't skip levels.",
      },
    ],
    "level": 3,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kb2fead6",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k752af68",
        "_type": "span",
        "marks": Array [],
        "text": "How these are represented in the same list is an interesting question.",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "ke2dc601",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k764d88f",
        "_type": "span",
        "marks": Array [],
        "text": "Another thing to note",
      },
    ],
    "markDefs": Array [],
    "style": "h4",
  },
  Object {
    "_key": "k8c6d45d",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k3c854a2",
        "_type": "span",
        "marks": Array [],
        "text": "Ordered lists are also a thing. I'm assuming they're implemented the same as bullet lists.",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kc40fbfc",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kc6e1ffd",
        "_type": "span",
        "marks": Array [],
        "text": "Cause that would make sense.",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k7b95f4b",
    "_type": "block",
    "children": Array [],
    "markDefs": Array [],
    "style": "blockquote",
  },
  Object {
    "_key": "k98232f4",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k70999b7",
        "_type": "span",
        "marks": Array [],
        "text": "Here's a reference to an entry as a block level thing:",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k91a81f4",
    "_ref": "beer-1572281",
    "_type": "reference",
  },
  Object {
    "_key": "k0389406",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kaa5d1ab",
        "_type": "span",
        "marks": Array [],
        "text": "And ",
      },
      Object {
        "_key": "k3784a81",
        "_type": "span",
        "marks": Array [],
        "text": " is an inline reference to one. Another option is to link to an ",
      },
      Object {
        "_key": "k1159efe",
        "_type": "span",
        "marks": Array [],
        "text": ".",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k31f29fc",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kcadbed3",
        "_type": "span",
        "marks": Array [],
        "text": "Here's an inline ",
      },
      Object {
        "_key": "k919d6e4",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "entry:",
      },
      Object {
        "_key": "k31d2227",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "which",
      },
      Object {
        "_key": "k412fffa",
        "_type": "span",
        "marks": Array [],
        "text": " is pretty cool.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "ke111e3a",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k5733cdd",
        "_type": "span",
        "marks": Array [],
        "text": "Who would have thought you could embed assets, huh?",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kf189343",
    "_ref": "3LJQayn0FiaYi8K2m6KMia",
    "_type": "reference",
  },
  Object {
    "_key": "k6b36e97",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kfc4aa99",
        "_type": "span",
        "marks": Array [],
        "text": "That's it for today? I think that's all the features it supports.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
]
`;

exports[`toPortableText supports all features 1`] = `
Array [
  Object {
    "_key": "ka3360a9",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kd9865f3",
        "_type": "span",
        "marks": Array [],
        "text": "Hello, everyone!",
      },
    ],
    "markDefs": Array [],
    "style": "h1",
  },
  Object {
    "_key": "kbf477f7",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k9ee64b5",
        "_type": "span",
        "marks": Array [],
        "text": "This is a little introduction. It's just a paragraph without any fancy stuff.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k3ec023b",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k77063cd",
        "_type": "span",
        "marks": Array [],
        "text": "Background",
      },
    ],
    "markDefs": Array [],
    "style": "h2",
  },
  Object {
    "_key": "kf72d303",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k1864115",
        "_type": "span",
        "marks": Array [],
        "text": "When representing text with formatting (such as ",
      },
      Object {
        "_key": "k972fe1e",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "bold",
      },
      Object {
        "_key": "kab7a70e",
        "_type": "span",
        "marks": Array [],
        "text": ", ",
      },
      Object {
        "_key": "ka074963",
        "_type": "span",
        "marks": Array [
          "em",
        ],
        "text": "italic",
      },
      Object {
        "_key": "kab7a70e",
        "_type": "span",
        "marks": Array [],
        "text": ", ",
      },
      Object {
        "_key": "kd3067d7",
        "_type": "span",
        "marks": Array [
          "underline",
        ],
        "text": "underline",
      },
      Object {
        "_key": "kbeb12c2",
        "_type": "span",
        "marks": Array [],
        "text": " or ",
      },
      Object {
        "_key": "ka1bcaba",
        "_type": "span",
        "marks": Array [
          "code",
        ],
        "text": "code",
      },
      Object {
        "_key": "ke530559",
        "_type": "span",
        "marks": Array [],
        "text": "), you want to do this in a ",
      },
      Object {
        "_key": "ka50426c",
        "_type": "span",
        "marks": Array [
          "k933711f",
        ],
        "text": "platform-agnostic",
      },
      Object {
        "_key": "kaebcdd6",
        "_type": "span",
        "marks": Array [],
        "text": " way, so you are able to render it on any device/surface.",
      },
    ],
    "markDefs": Array [
      Object {
        "_key": "k933711f",
        "_type": "link",
        "href": "https://www.techopedia.com/definition/23666/platform-agnostic",
      },
    ],
    "style": "normal",
  },
  Object {
    "_key": "k1380a38",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k4f7db8f",
        "_type": "span",
        "marks": Array [],
        "text": "The situation gets more complex when you want to combine ",
      },
      Object {
        "_key": "kafe874d",
        "_type": "span",
        "marks": Array [
          "strong",
          "em",
          "k1c64ec3",
        ],
        "text": "multiple",
      },
      Object {
        "_key": "kccbade6",
        "_type": "span",
        "marks": Array [],
        "text": " formatting options, or when several \\"marks\\" are split ",
      },
      Object {
        "_key": "k0b7a9d5",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "across ",
      },
      Object {
        "_key": "k88a89d3",
        "_type": "span",
        "marks": Array [
          "strong",
          "underline",
        ],
        "text": "multiple",
      },
      Object {
        "_key": "k9e33411",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": " words",
      },
      Object {
        "_key": "k1159efe",
        "_type": "span",
        "marks": Array [],
        "text": ".",
      },
    ],
    "markDefs": Array [
      Object {
        "_key": "k1c64ec3",
        "_type": "link",
        "href": "https://multiple.com/",
      },
    ],
    "style": "normal",
  },
  Object {
    "_key": "k9c4bf48",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k6494432",
        "_type": "span",
        "marks": Array [],
        "text": "Simple lists are supported, of course.",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kdff0bd7",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kb19e8e0",
        "_type": "span",
        "marks": Array [],
        "text": "Indented ones as well",
      },
    ],
    "level": 2,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k4606371",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k99acc14",
        "_type": "span",
        "marks": Array [],
        "text": "But you can't skip levels.",
      },
    ],
    "level": 3,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kb2fead6",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k752af68",
        "_type": "span",
        "marks": Array [],
        "text": "How these are represented in the same list is an interesting question.",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "ke2dc601",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k764d88f",
        "_type": "span",
        "marks": Array [],
        "text": "Another thing to note",
      },
    ],
    "markDefs": Array [],
    "style": "h4",
  },
  Object {
    "_key": "k8c6d45d",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k3c854a2",
        "_type": "span",
        "marks": Array [],
        "text": "Ordered lists are also a thing. I'm assuming they're implemented the same as bullet lists.",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kc40fbfc",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kc6e1ffd",
        "_type": "span",
        "marks": Array [],
        "text": "Cause that would make sense.",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k7b95f4b",
    "_type": "block",
    "children": Array [],
    "markDefs": Array [],
    "style": "blockquote",
  },
  Object {
    "_key": "k98232f4",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k70999b7",
        "_type": "span",
        "marks": Array [],
        "text": "Here's a reference to an entry as a block level thing:",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k91a81f4",
    "_ref": "beer-1572281",
    "_type": "reference",
  },
  Object {
    "_key": "k0389406",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kaa5d1ab",
        "_type": "span",
        "marks": Array [],
        "text": "And ",
      },
      Object {
        "_key": "k3784a81",
        "_type": "span",
        "marks": Array [],
        "text": " is an inline reference to one. Another option is to link to an ",
      },
      Object {
        "_key": "k1159efe",
        "_type": "span",
        "marks": Array [],
        "text": ".",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "k31f29fc",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kcadbed3",
        "_type": "span",
        "marks": Array [],
        "text": "Here's an inline ",
      },
      Object {
        "_key": "k919d6e4",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "entry:",
      },
      Object {
        "_key": "k31d2227",
        "_type": "span",
        "marks": Array [
          "strong",
        ],
        "text": "which",
      },
      Object {
        "_key": "k412fffa",
        "_type": "span",
        "marks": Array [],
        "text": " is pretty cool.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "ke111e3a",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "k5733cdd",
        "_type": "span",
        "marks": Array [],
        "text": "Who would have thought you could embed assets, huh?",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
  Object {
    "_key": "kf189343",
    "_ref": "3LJQayn0FiaYi8K2m6KMia",
    "_type": "reference",
  },
  Object {
    "_key": "k6b36e97",
    "_type": "block",
    "children": Array [
      Object {
        "_key": "kfc4aa99",
        "_type": "span",
        "marks": Array [],
        "text": "That's it for today? I think that's all the features it supports.",
      },
    ],
    "markDefs": Array [],
    "style": "normal",
  },
]
`;
